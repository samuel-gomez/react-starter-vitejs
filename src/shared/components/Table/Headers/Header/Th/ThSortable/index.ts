export { default } from './ThSortable.container';
export type { TThSortableContainer, Torder, TsetSort } from './ThSortable.container';
export { default as ThSortable } from './ThSortable';
export type { TThSortable } from './ThSortable';
