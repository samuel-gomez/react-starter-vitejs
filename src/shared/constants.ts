import packageJson from '../../package.json';

/** ***********************************************************************
 * FORM CONSTANTS
 ************************************************************************ */
export const MSG_REQUIRED = 'Le champ est obligatoire';

export const MSG_LENGTH_3 = 'Le champ doit faire au moins 3 caractères';
export const MSG_FORMAT = 'Format invalide';
export const MSG_ERROR_FORMAT = 'Format de fichier incorrect';

export const DEFAULT_OPTION_LABEL = '- Sélectionner -';
export const DEFAULT_OPTION = {
  value: '',
  label: DEFAULT_OPTION_LABEL,
};

// patterns
export const patternEmail = /^[\w-+]+@([\w-+]+\.)+[\w-]{2,}$/;
export const patternPhoneNumber = /^[+0][\d ]*$/;
export const patternZipCode = /^\d{5}$/;
export const patternSiret = /^\d{14}$/;
export const patternSiren = /^\d{9}$/;
export const patternSiretSiren = /^\d{9,14}$/;

export const phoneNumberRule = {
  required: true,
  minLength: 4,
  maxLength: 20,
  pattern: patternPhoneNumber,
};

export const emailRule = {
  required: true,
  minLength: 6,
  maxLength: 60,
  pattern: patternEmail,
};

/** ***********************************************************************
 * API STATUS CONSTANTS
 ************************************************************************ */
export const STATUS_HTTP = {
  SUCCESS: 200,
  CREATED: 201,
  BAD_REQUEST: 400,
  UNAUTHORIZE: 401,
  FORBIDDEN: 403,
  NOTFOUND: 404,
  SERVER_ERROR: 500,
};

export const STATUS_API = {
  SUCCESS: STATUS_HTTP.SUCCESS,
  CREATED: STATUS_HTTP.CREATED,
  WARNING: 40,
  ERROR: 50,
};

/** ***********************************************************************
 * API URL
 ************************************************************************ */
export const API_URL = {
  BASE: 'base',
  GITHUB: 'github',
};

export const STATUS_HTTP_MESSAGES = {
  [STATUS_HTTP.SUCCESS]: 'Succès: La requête a réussi',
  [STATUS_HTTP.CREATED]: 'Succès: Création réussie',
  [STATUS_HTTP.BAD_REQUEST]: 'Erreur: La syntaxe de la requête est erronée',
  [STATUS_HTTP.UNAUTHORIZE]: 'Erreur: Une authentification est nécessaire pour accéder à la ressource',
  [STATUS_HTTP.FORBIDDEN]: 'Erreur: Accès non autorisé',
  [STATUS_HTTP.NOTFOUND]: 'Erreur: Elément non trouvé',
  [STATUS_HTTP.SERVER_ERROR]: 'Erreur: Problème technique ! Contacter votre support',
};

/** ***********************************************************************
 * USER CONSTANTS
 ************************************************************************ */

export const ADMIN = 'Admin';
export const READER = 'Reader';
export const WRITER = 'Writer';
export const USER = 'User';
export const ALL = '';

export const PROFILS = [ADMIN, READER, WRITER, USER, ALL];

/** ***********************************************************************
 * COMMONS CONSTANTS
 ************************************************************************ */

export const MODIFIER_CLASS = {
  success: 'success',
  disabled: 'disabled',
};

/** ***********************************************************************
 * VERSION CONSTANT
 ************************************************************************ */

const VERSION_TOOLKIT = 'master';

/** ***********************************************************************
 * DEMO CONSTANTS
 ************************************************************************ */

export const STORYBOOK = 'https://axafrance.github.io/react-toolkit/latest/storybook/?path=/story/';
export const DESIGN_SYSTEM = 'https://axafrance.github.io/design-system/';
export const GITHUB = `https://github.com/AxaGuilDEv/react-toolkit/tree/v${packageJson.dependencies['@axa-fr/react-toolkit-all']}/packages/`;
export const GITHUB_API = `AxaGuilDEv/react-toolkit/${VERSION_TOOLKIT}/packages/`;
