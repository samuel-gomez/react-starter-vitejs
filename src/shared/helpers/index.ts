export { formatDate, isValidDate, setDate } from './formatDate';
export { default as isEmptyOrNull, isEmptyArray, isEmptyObject } from './isEmptyOrNull';
export { default as mergeObj } from './mergeObj';
export { default as setAnomalyEmptyItems } from './setAnomalyEmptyItems';
export { default as setConfirmClassModifier } from './setConfirmClassModifier';
export { default as emptyFunction } from './emptyFunction';
export { default as setRequiredFieldRule, setRequiredMessage } from './setRequiredFieldRule';
