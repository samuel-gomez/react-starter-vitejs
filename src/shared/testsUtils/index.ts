export * from './customRender';
export { default as emptyFunction } from '../helpers/emptyFunction';
export { default as MOCK_API_URL } from './constants';
export { default as renderWithContainer } from './renderWithContainer';
