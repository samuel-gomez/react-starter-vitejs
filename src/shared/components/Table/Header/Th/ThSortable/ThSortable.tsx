import { ReactNode } from 'react';
import { withClassDefault, withClassModifier, WithClassModifierOptions, compose, identity } from '@axa-fr/react-toolkit-core';
import { NONE, ASCENDING, DESCENDING } from 'shared/components/Table/constants';
import type { Torder } from './ThSortable.container';

export const orderIcons = (order: Torder = NONE) =>
  ({
    [NONE]: 'sorting',
    [ASCENDING]: 'arrow-xs-up',
    [DESCENDING]: 'arrow-xs-down',
  }[order]);

export type TSortingIcon = {
  orderIconsFn?: typeof orderIcons;
  order?: Torder;
};
export const SortingIcon = ({ order, orderIconsFn = orderIcons }: TSortingIcon) => (
  <span className={`af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-${orderIconsFn(order)}`} />
);

export type TThSortable = {
  className?: string;
  children?: ReactNode;
  sort: () => void;
  order: Torder;
} & WithClassModifierOptions;

const DEFAULT_CLASSNAME = 'af-table__th';

const ThSortable = ({ className, children, sort, order }: TThSortable) => (
  <th role="button" onClick={sort} className={className}>
    {children}
    <SortingIcon order={order} />
  </th>
);

const enhance = compose(identity<TThSortable>(), withClassDefault(DEFAULT_CLASSNAME), withClassModifier());

const Enhanced = enhance(ThSortable);
Enhanced.displayName = ThSortable.name;

export default Enhanced;
