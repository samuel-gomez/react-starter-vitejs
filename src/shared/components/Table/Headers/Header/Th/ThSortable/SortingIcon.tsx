import { NONE, ASCENDING, DESCENDING } from 'shared/components/Table/constants';
import type { Torder } from './ThSortable.container';

export const orderIcons = (order: Torder = NONE) =>
  ({
    [NONE]: 'sorting',
    [ASCENDING]: 'arrow-xs-up',
    [DESCENDING]: 'arrow-xs-down',
  })[order];

export type TSortingIcon = {
  orderIconsFn?: typeof orderIcons;
  order?: Torder;
};

const SortingIcon = ({ order, orderIconsFn = orderIcons }: TSortingIcon) => (
  <span className={`af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-${orderIconsFn(order)}`} />
);

export default SortingIcon;
