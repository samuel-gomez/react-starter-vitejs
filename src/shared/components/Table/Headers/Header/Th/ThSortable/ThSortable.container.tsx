import { ASCENDING, DESCENDING, NONE } from 'shared/components/Table/constants';
import ThSortable, { type TThSortable } from './ThSortable';

export type Torder = typeof ASCENDING | typeof DESCENDING | typeof NONE;
export const toggleOrder = (order: Torder) => (order === ASCENDING ? DESCENDING : ASCENDING);

type TsetOrder = {
  toggleOrderFn?: typeof toggleOrder;
  order: Torder;
  field: string;
  prevField: string;
};

export const setOrder = ({ field, prevField, order, toggleOrderFn = toggleOrder }: TsetOrder) =>
  prevField !== field ? ASCENDING : toggleOrderFn(order);

type Tsorting = {
  field: string;
  order: Torder;
};

type TtoggleSorting = {
  setOrderFn?: typeof setOrder;
  field: string;
  sorting?: Tsorting;
};

export const toggleSorting = ({ field, sorting: { field: prevField, order } = { field, order: NONE }, setOrderFn = setOrder }: TtoggleSorting) => ({
  field,
  order: setOrderFn({ field, prevField, order }) as Torder,
});

type TsetClassModifierActive = {
  classModifier?: string;
  isActive?: boolean;
};

export const setClassModifierActive = ({ classModifier = '', isActive = false }: TsetClassModifierActive) =>
  isActive ? `${classModifier} active` : classModifier;

export type TsetSort = TtoggleSorting & {
  onSort: (arg: ReturnType<typeof toggleSorting>) => void;
};

export const setSort = ({ field, sorting, onSort, toggleSortingFn = toggleSorting }: TsetSort & { toggleSortingFn?: typeof toggleSorting }) => {
  const sortBy = toggleSortingFn({ field, sorting });
  const sort = () => onSort(sortBy);
  return sort;
};

export type TThSortableContainer = Omit<TThSortable, 'sort' | 'sortingIcon' | 'order'> & {
  classModifier?: string;
  ThSortableCmpt?: typeof ThSortable;
  setClassModifierActiveFn?: typeof setClassModifierActive;
  setSortFn?: typeof setSort;
} & TsetSort;

const ThSortableContainer = ({
  field,
  onSort,
  sorting = { field, order: NONE },
  classModifier = 'sortable',
  setClassModifierActiveFn = setClassModifierActive,
  ThSortableCmpt = ThSortable,
  setSortFn = setSort,
  ...rest
}: TThSortableContainer) => {
  const { order, field: fieldActive } = sorting;
  const isActive = fieldActive === field;
  const sort = setSortFn({ field, sorting, onSort });
  const newClassModifier = setClassModifierActiveFn({ classModifier, isActive });

  return <ThSortableCmpt {...rest} classModifier={newClassModifier} sort={sort} order={isActive ? order : NONE} />;
};

export default ThSortableContainer;
