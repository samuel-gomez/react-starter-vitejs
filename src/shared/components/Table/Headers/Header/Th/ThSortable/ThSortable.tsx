import { Table as TableTk } from '@axa-fr/react-toolkit-all';
import { type ComponentPropsWithoutRef } from 'react';
import SortingIcon from './SortingIcon';
import type { Torder } from './ThSortable.container';

export type TThSortable = ComponentPropsWithoutRef<typeof TableTk.Th> & {
  sort: () => void;
  order: Torder;
};

const ThSortable = ({ className, children, sort, order, ...thSortableProps }: TThSortable) => (
  <TableTk.Th role="button" onClick={sort} {...thSortableProps}>
    {children}
    <SortingIcon order={order} />
  </TableTk.Th>
);

export default ThSortable;
