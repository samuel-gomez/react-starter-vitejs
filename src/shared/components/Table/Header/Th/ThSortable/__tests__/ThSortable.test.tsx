import { describe, it, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import { renderWithContainer } from 'shared/testsUtils';
import ThSortable, { SortingIcon, orderIcons } from '../ThSortable';

const defaultProps = {
  sort: vi.fn(),
};
const container = document.createElement('tr');

describe('ThSortable', () => {
  it.each`
    order     | children     | classModifier
    ${''}     | ${undefined} | ${undefined}
    ${'NONE'} | ${undefined} | ${undefined}
    ${'NONE'} | ${'child'}   | ${undefined}
    ${'NONE'} | ${'child'}   | ${'variant'}
    ${1}      | ${'child'}   | ${'variant'}
    ${-1}     | ${'child'}   | ${'variant'}
  `(
    'Should render <ThSortable/> when order: $order, children: $children, className: $className, classModifier: $classModifier',
    ({ children, ...rest }) => {
      const { baseElement } = renderWithContainer(
        <ThSortable {...defaultProps} {...rest}>
          {children}
        </ThSortable>,
        container,
      );

      expect(baseElement).toMatchSnapshot();
    },
  );
});

describe('orderIcons', () => {
  it.each`
    order        | expected
    ${undefined} | ${'sorting'}
    ${'NONE'}    | ${'sorting'}
    ${1}         | ${'arrow-xs-up'}
    ${-1}        | ${'arrow-xs-down'}
  `('Should return $expected when orderIcons have been called with order: $order', ({ order, expected }) => {
    const result = orderIcons(order);

    expect(result).toEqual(expected);
  });
});

describe('SortingIcon', () => {
  it('Should render SortingIcon with order equal NONE', () => {
    const { asFragment } = render(<SortingIcon />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render SortingIcon with order equal 1', () => {
    const { asFragment } = render(<SortingIcon order={1} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Should render SortingIcon with order equal -1', () => {
    const { asFragment } = render(<SortingIcon order={-1} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
