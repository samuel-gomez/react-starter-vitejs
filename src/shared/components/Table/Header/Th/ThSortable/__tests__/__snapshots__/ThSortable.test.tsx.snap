// Vitest Snapshot v1

exports[`SortingIcon > Should render SortingIcon with order equal -1 1`] = `
<DocumentFragment>
  <span
    class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-arrow-xs-down"
  />
</DocumentFragment>
`;

exports[`SortingIcon > Should render SortingIcon with order equal 1 1`] = `
<DocumentFragment>
  <span
    class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-arrow-xs-up"
  />
</DocumentFragment>
`;

exports[`SortingIcon > Should render SortingIcon with order equal NONE 1`] = `
<DocumentFragment>
  <span
    class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
  />
</DocumentFragment>
`;

exports[`ThSortable > Should render <ThSortable/> when order: '', children: undefined, className: undefined, classModifier: undefined 1`] = `
<tr>
  <th
    class="af-table__th"
    role="button"
  >
    <span
      class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-undefined"
    />
  </th>
</tr>
`;

exports[`ThSortable > Should render <ThSortable/> when order: 'NONE', children: 'child', className: undefined, classModifier: 'variant' 1`] = `
<tr>
  <th
    class="af-table__th af-table__th--variant"
    role="button"
  >
    child
    <span
      class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
    />
  </th>
</tr>
`;

exports[`ThSortable > Should render <ThSortable/> when order: 'NONE', children: 'child', className: undefined, classModifier: undefined 1`] = `
<tr>
  <th
    class="af-table__th"
    role="button"
  >
    child
    <span
      class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
    />
  </th>
</tr>
`;

exports[`ThSortable > Should render <ThSortable/> when order: 'NONE', children: undefined, className: undefined, classModifier: undefined 1`] = `
<tr>
  <th
    class="af-table__th"
    role="button"
  >
    <span
      class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
    />
  </th>
</tr>
`;

exports[`ThSortable > Should render <ThSortable/> when order: -1, children: 'child', className: undefined, classModifier: 'variant' 1`] = `
<tr>
  <th
    class="af-table__th af-table__th--variant"
    role="button"
  >
    child
    <span
      class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-arrow-xs-down"
    />
  </th>
</tr>
`;

exports[`ThSortable > Should render <ThSortable/> when order: 1, children: 'child', className: undefined, classModifier: 'variant' 1`] = `
<tr>
  <th
    class="af-table__th af-table__th--variant"
    role="button"
  >
    child
    <span
      class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-arrow-xs-up"
    />
  </th>
</tr>
`;
